sqrl-planner/src/main/java/com/datasqrl/io/schema/avro/AvroToRelDataTypeConverter.java

```diff
@@ -21,92 +26,130 @@ public class AvroToRelDataTypeConverter {
 
   private final TypeFactory typeFactory;
   private final ErrorCollector errors;
+  private final Map<Schema, RelDataType> typeCache;
 
   public AvroToRelDataTypeConverter(ErrorCollector errors) {
-    this(TypeFactory.getTypeFactory(), errors);
+    this(TypeFactory.getTypeFactory(), errors, new IdentityHashMap<>());
   }
 
   public RelDataType convert(Schema schema) {
     return convertAvroSchemaToCalciteType(schema, NamePath.ROOT);
   }
 
   private RelDataType convertAvroSchemaToCalciteType(Schema schema, NamePath path) {
-    RelDataType relType;
+    // Check if the schema has already been processed
+    if (typeCache.containsKey(schema)) {
+      throw errors.exception(ErrorCode.SCHEMA_ERROR, "Recursive schema's not yet supported: %s", path);
+    }
+
     switch (schema.getType()) {
       case UNION:
```

Comment:
To make the code a bit cleaner and DRYer, use the new java switch expression, i.e. something like

```relType = switch (schema.getType()) {
}
typeCache.add(schema,relType);
return relType;
```